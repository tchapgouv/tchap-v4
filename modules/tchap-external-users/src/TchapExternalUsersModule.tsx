/*
Copyright 2022 New Vector Ltd. t/a Element

Licensed under the Apache License, Version 2.0 (the "License");
you may not use this file except in compliance with the License.
You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

Unless required by applicable law or agreed to in writing, software
distributed under the License is distributed on an "AS IS" BASIS,
WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
See the License for the specific language governing permissions and
limitations under the License.
*/

import { RuntimeModule } from "@matrix-org/react-sdk-module-api/lib/RuntimeModule";
import { ModuleApi } from "@matrix-org/react-sdk-module-api/lib/ModuleApi";
import { CantAddDialog  } from "./components/CantAddDialog";
import { AddRoomListener, ExternalUsersLifecycle } from "@matrix-org/react-sdk-module-api/lib/tchap/lifecycles/ExternalUsers";
import { DialogProps } from "@matrix-org/react-sdk-module-api/lib/components/DialogContent";

import React from "react";

export default class TchapExternalUsersModule extends RuntimeModule {
    private domain: String;

    public constructor(moduleApi: ModuleApi) {
        super(moduleApi);

        // const rawConfig = this.moduleApi.getConfigValue(
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_NAMESPACE,
        //     TCHAP_EXTERNAL_USERS_MODULE_CONFIG_KEY,
        //   );
      
        //   assertValidGuestModuleConfig(rawConfig);
        //   this.config = rawConfig;

        this.on(ExternalUsersLifecycle.AddRoom, this.onAddRoom);
    }

    protected onAddRoom: AddRoomListener = async () => {
        await this.moduleApi.openDialog<null, DialogProps, CantAddDialog>("test", (props, ref) => <CantAddDialog ref={ref} {...props} />);
    };
}
