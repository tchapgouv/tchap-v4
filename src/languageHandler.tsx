/*
Copyright 2024 New Vector Ltd.
Copyright 2023 The Matrix.org Foundation C.I.C.

SPDX-License-Identifier: AGPL-3.0-only OR GPL-3.0-only
Please see LICENSE files in the repository root for full details.
*/

import React from "react";
import {
    IVariables,
    TranslatedString,
    TranslationKey as ReactTranslationKey,
    // eslint-disable-next-line camelcase
    _t as react_t,
    // eslint-disable-next-line camelcase
    _td as react_td,
    // eslint-disable-next-line camelcase
    _tDom as react_tDom,
    Tags,
    UserFriendlyError as ReactUserFriendlyError,
    ErrorOptions,
} from "matrix-react-sdk/src/languageHandler";
import { Leaves } from "matrix-react-sdk/src/@types/common";

import type ReactEN from "matrix-react-sdk/src/i18n/strings/en_EN.json";
import type EN from "./i18n/strings/en_EN.json";

/**
 * This module wraps languageHandler in the matrix-react-sdk and adds type casts to include translations
 * which we know will be injected by webpack.
 */

export type TranslationKey = Leaves<typeof EN & typeof ReactEN, "|", string | { other: string }, 4>;

export class UserFriendlyError extends ReactUserFriendlyError {
    public constructor(message: TranslationKey, substitutionVariablesAndCause?: IVariables & ErrorOptions) {
        super(message as ReactTranslationKey, substitutionVariablesAndCause);
    }
}

export function _td(s: TranslationKey): TranslationKey {
    return react_td(s as ReactTranslationKey);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _t(text: TranslationKey, variables?: IVariables): string;
export function _t(text: TranslationKey, variables: IVariables | undefined, tags: Tags): React.ReactNode;
export function _t(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_t(text as ReactTranslationKey, variables, tags!);
}

// eslint-next-line @typescript-eslint/naming-convention
export function _tDom(text: TranslationKey, variables?: IVariables): TranslatedString;
export function _tDom(text: TranslationKey, variables: IVariables, tags: Tags): React.ReactNode;
export function _tDom(text: TranslationKey, variables?: IVariables, tags?: Tags): TranslatedString {
    return react_tDom(text as ReactTranslationKey, variables!, tags!);
}
