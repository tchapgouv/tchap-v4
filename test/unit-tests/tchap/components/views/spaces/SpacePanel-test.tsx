import React from "react";
import { render, cleanup } from "@testing-library/react";

import { stubClient } from "~matrix-react-sdk/test/test-utils";
import { MatrixClientPeg } from "~matrix-react-sdk/src/MatrixClientPeg";
import { wrapInMatrixClientContext, wrapInSdkContext } from "~matrix-react-sdk/test/test-utils";
import { SdkContextClass } from "~matrix-react-sdk/src/contexts/SDKContext";
import UnwrappedSpacePanel from "~matrix-react-sdk/src/components/views/spaces/SpacePanel";
import SdkConfig, { ConfigOptions } from "~tchap-web/linked-dependencies/matrix-react-sdk/src/SdkConfig";

describe("<SpacePanel />", () => {
    const SpacePanel = wrapInSdkContext(wrapInMatrixClientContext(UnwrappedSpacePanel), SdkContextClass.instance);
    const featureName: string = "feature_thread";
    const homeserverName: string = "my.home.server";

    const addHomeserverToMockConfig = (homeservers: string[]) => {
        // mock SdkConfig.get("tchap_features")
        const config: ConfigOptions = { tchap_features: {} };
        config.tchap_features[featureName] = homeservers;
        SdkConfig.put(config);
    };

    const renderSpacePanel = () => render(<SpacePanel />);

    beforeEach(() => {
        stubClient();
        MatrixClientPeg.getHomeserverName = () => homeserverName;
    });

    afterEach(() => {
        cleanup();
    });

    it("returns true when the the homeserver include thread feature", () => {
        addHomeserverToMockConfig([homeserverName]);
        const { container } = renderSpacePanel();

        expect(container.getElementsByClassName("mx_ThreadsActivityCentre_container").length).toBe(1);
    });

    it("returns false when the the homeserver doesnt include thread feature", async () => {
        addHomeserverToMockConfig(["other.homeserver"]);
        const { container } = renderSpacePanel();

        expect(container.getElementsByClassName("mx_ThreadsActivityCentre_container").length).toBe(0);
    });
});
