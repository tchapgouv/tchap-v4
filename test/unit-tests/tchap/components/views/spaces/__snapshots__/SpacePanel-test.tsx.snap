// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<SpacePanel /> should render as expected 1`] = `
<div>
  <nav
    aria-label="Spaces"
    class="mx_SpacePanel collapsed"
  >
    <div
      class="mx_UserMenu"
    >
      <div
        aria-expanded="false"
        aria-haspopup="true"
        aria-label="User menu"
        class="mx_AccessibleButton mx_UserMenu_contextMenuButton"
        role="button"
        tabindex="0"
      >
        <div
          class="mx_UserMenu_userAvatar"
        >
          <span
            class="_avatar_mcap2_17 mx_BaseAvatar mx_UserMenu_userAvatar_BaseAvatar _avatar-imageless_mcap2_61"
            data-color="2"
            data-testid="avatar-img"
            data-type="round"
            role="presentation"
            style="--cpd-avatar-size: 32px;"
          >
            u
          </span>
        </div>
      </div>
      <div
        aria-label="Expand"
        class="mx_AccessibleButton mx_SpacePanel_toggleCollapse"
        role="button"
        tabindex="0"
      />
    </div>
    <ul
      aria-label="Spaces"
      class="mx_AutoHideScrollbar mx_SpaceTreeLevel"
      data-rbd-droppable-context-id="0"
      data-rbd-droppable-id="top-level-spaces"
      role="tree"
      tabindex="-1"
    >
      <li
        aria-selected="false"
        class="mx_SpaceItem mx_SpaceItem_new collapsed"
        role="treeitem"
      >
        <div
          aria-label="Create a space"
          class="mx_AccessibleButton mx_SpaceButton mx_SpaceButton_new mx_SpaceButton_narrow"
          data-testid="create-space-button"
          role="button"
          tabindex="0"
        >
          <div
            class="mx_SpaceButton_selectionWrapper"
          >
            <div
              class="mx_SpaceButton_avatarWrapper"
            >
              <div
                class="mx_SpaceButton_avatarPlaceholder"
              >
                <div
                  class="mx_SpaceButton_icon"
                />
              </div>
            </div>
          </div>
        </div>
      </li>
    </ul>
    <div
      aria-expanded="false"
      aria-label="Help"
      class="mx_AccessibleButton mx_QuickSettingsButton tc_sidebar_quick_faq"
      role="button"
      tabindex="0"
    />
    <div>
      Mocked ChildComponent with text: 
    </div>
  </nav>
</div>
`;
